pkg/pillar/cmd/volumemgr/handlevolume.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package volumemgr
+
+import (
+	"time"
+
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	log "github.com/sirupsen/logrus"
+)
+
+func handleVolumeCreate(ctxArg interface{}, key string,
+	configArg interface{}) {
+
+	log.Infof("handleVolumeCreate(%s)", key)
+	config := configArg.(types.VolumeConfig)
+	ctx := ctxArg.(*volumemgrContext)
+	updateVolume(ctx, config)
+	log.Infof("handleVolumeCreate(%s) Done", key)
+}
+
+func handleVolumeModify(ctxArg interface{}, key string,
+	configArg interface{}) {
+
+	log.Infof("handleVolumeModify(%s)", key)
+	config := configArg.(types.VolumeConfig)
+	ctx := ctxArg.(*volumemgrContext)
+	updateVolume(ctx, config)
+	log.Infof("handleVolumeModify(%s) Done", key)
+}
+
+func handleVolumeDelete(ctxArg interface{}, key string,
+	configArg interface{}) {
+
+	log.Infof("handleVolumeDelete(%s)", key)
+	config := configArg.(types.VolumeConfig)
+	ctx := ctxArg.(*volumemgrContext)
+	deleteVolume(ctx, config)
+	log.Infof("handleVolumeDelete(%s) Done", key)
+}
+
+func publishVolumeStatus(ctx *volumemgrContext,
+	status *types.VolumeStatus) {
+
+	key := status.Key()
+	log.Debugf("publishVolumeStatus(%s)", key)
+	pub := ctx.pubVolumeStatus
+	pub.Publish(key, *status)
+	log.Debugf("publishVolumeStatus(%s) Done", key)
+}
+
+func unpublishVolumeStatus(ctx *volumemgrContext,
+	status *types.VolumeStatus) {
+
+	key := status.Key()
+	log.Debugf("unpublishVolumeStatus(%s)", key)
+	pub := ctx.pubVolumeStatus
+	c, _ := pub.Get(key)
+	if c == nil {
+		log.Errorf("unpublishVolumeStatus(%s) not found", key)
+		return
+	}
+	pub.Unpublish(key)
+	log.Debugf("unpublishVolumeStatus(%s) Done", key)
+}
+
+func lookupVolumeStatus(ctx *volumemgrContext,
+	key string) *types.VolumeStatus {
+
+	log.Infof("lookupVolumeStatus(%s)", key)
+	pub := ctx.pubVolumeStatus
+	c, _ := pub.Get(key)
+	if c == nil {
+		log.Infof("lookupVolumeStatus(%s) not found", key)
+		return nil
+	}
+	status := c.(types.VolumeStatus)
+	log.Infof("lookupVolumeStatus(%s) Done", key)
+	return &status
+}
+
+func updateVolume(ctx *volumemgrContext,
```

Comment:
When handleCreate calls this, don't you need to increment the refcount? Otherwise refcount stays at zero.

Reply:
No, refcount will not be incremented in handleCreate call in volumemgr. Refount will be incremented by the zedagent based on the number of volume refs available in the app instance config. Today, no app instance is using these new volumes so refcount set to zero makes sense.

Reply:
Added. Incrementing refcount in volume config published from the zedagent.